#' For anyone reviewing the code below, the following is a small style guide 
#' outlining the various formats for the code. 
#' 
#' Names with "_": objects, inlucding data.frames, GRanges, vectors, ...
#' Names in caMel format: functions or components of objects (i.e. columns 
#' within a data.frame).
#' Names with ".": arguments / options for functions

options(stringsAsFactors = FALSE, scipen = 99, width = 180)


# Set up and gather command line arguments ----
parser <- argparse::ArgumentParser(
  description = "Generate an iGUIDE report for input evaluation data.",
  usage = paste(
    "Rscript generate_IGUIDE_report.R <eval.rds> -o <output>",
    "[-h/--help, -v/--version] [optional args]"
  )
)

parser$add_argument(
  "eval", nargs = 1, type = "character",
  help = paste(
    "Evaluation dataset, in rds format. Can be generated by the",
    "'iguide eval' subcommand."
  )
)

parser$add_argument(
  "-o", "--output", nargs = 1, type = "character", required = TRUE,
  help = "Output report file, extension not required."
)

parser$add_argument(
  "-b", "--tables", action = "store_true",
  help = "Generate tables along with output report (csv formats)."
)

parser$add_argument(
  "-f", "--figures", action = "store_true",
  help = "Generate figures along with output report (pdf and png formats)."
)

parser$add_argument(
  "-d", "--data", action = "store_true",
  help = "Data to generate the report will be saved as an R image with output."
)

parser$add_argument(
  "-t", "--format", nargs = 1, type = "character", default = "html",
  help = "Output format for report. Either 'pdf' or 'html' (default)."
)

parser$add_argument(
  "-g", "--graphic", action = "store_true",
  help = "Includes an opening graphic on the report."
)

parser$add_argument(
  "--template", nargs = 1, type = "character", 
  default = "tools/rscripts/report_templates/iGUIDE_report_template.Rmd",
  help = "File path to standard or custom iGUIDE report template."
)

parser$add_argument(
  "--iguide_dir", nargs = 1, type = "character", default = "IGUIDE_DIR",
  help = "iGUIDE install directory path, do not change for normal applications."
)


args <- parser$parse_args(commandArgs(trailingOnly = TRUE))

if( !dir.exists(args$iguide_dir) ){
  root_dir <- Sys.getenv(args$iguide_dir)
}else{
  root_dir <- args$iguide_dir
}

if( !dir.exists(root_dir) ){
  stop(paste0("\n  Cannot find install path to iGUIDE: ", root_dir, ".\n"))
}else{
  args$iguide_dir <- root_dir
}

report_formats <- c("html" = "html_document", "pdf" = "pdf_document")

if( !args$format %in% names(report_formats) ){
  stop("Please input either 'html' or 'pdf' for format.\n",
       "Other formats not supported.")
}

output_format <- report_formats[args$format]

## Resolve template file path.
if( file.exists(file.path(root_dir, args$template)) ){
  
  template_path <- normalizePath(file.path(root_dir, args$template))
  
}else if( file.exists(file.path(args$template)) ){
  
  template_path <- normalizePath(file.path(args$template))
  
}else{
  
  stop("\nCannot find template file: ", args$template, ".\n")
  
}


## Construct input table and print to terminal
input_table <- data.frame(
  "Variables" = paste0(names(args), " :"), 
  "Values" = sapply(seq_along(args), function(i){
    paste(args[[i]], collapse = ", ")
  })
)

input_table <- input_table[
  match(
    c(
      "eval :", "output :", "tables :", "figures :", "data :", "graphic :", 
      "format :", "template :", "iguide_dir :"
    ),
    input_table$Variables),
]

cat("\niGUIDE Report Inputs:\n")

print(
  data.frame(input_table),
  right = FALSE, 
  row.names = FALSE
)


# Load dependancies ----
cat("\nLoading dependencies.\n")

add_packs <- c("magrittr", "knitr", "iguideSupport")

add_packs_loaded <- suppressMessages(
  sapply(add_packs, require, character.only = TRUE)
)

if( !all(add_packs_loaded) ){
  
  print(
    data.frame(
      "R-Packages" = names(add_packs_loaded), 
      "Loaded" = add_packs_loaded
    ), 
    right = FALSE,
    row.names = FALSE
  )
  
  stop("Check dependancies.\n")
  
}


# Import metadata and consolidate into report objects ----
cat("Importing evaluation data...\n")

if( file.exists(file.path(root_dir, args$eval)) ){
  
  eval_path <- normalizePath(file.path(root_dir, args$eval))
  
}else if( file.exists(file.path(args$eval)) ){
  
  eval_path <- normalizePath(file.path(args$eval))
  
}else{
  
  stop("\n  Cannot find evaluation dataset: ", args$eval, "\n")
  
}

eval_data <- readRDS(eval_path)

## Configuration
configs <- eval_data$params$configs

## Load reference genome 
if( grepl(".fa", unique(sapply(configs, "[[", "Ref_Genome"))) ){
  
  if( !(
    file.exists(
      file.path(root_dir, unique(sapply(configs, "[[", "Ref_Genome")))
    ) | file.exists(unique(sapply(configs, "[[", "Ref_Genome")))
  ) ){
    stop("Specified reference genome file not found.")
  }
  
  ref_file_type <- ifelse(
    grepl(".fastq", unique(sapply(configs, "[[", "Ref_Genome"))), 
    "fastq", 
    "fasta"
  )
  
  if( file.exists(
    file.path(root_dir, unique(sapply(configs, "[[", "Ref_Genome"))) 
    ) ){

    ref_genome <- Biostrings::readDNAStringSet(
      filepath = file.path(
        root_dir, unique(sapply(configs, "[[", "Ref_Genome"))
      ),
      format = ref_file_type
    )
    
  }else{
    
    ref_genome <- Biostrings::readDNAStringSet(
      filepath = unique(sapply(configs, "[[", "Ref_Genome")), 
      format = ref_file_type
    )
  }
  
}else{
  
  ref_genome <- unique(sapply(configs, "[[", "Ref_Genome"))
  
  genome <- grep(
    pattern = ref_genome, 
    x = unique(BSgenome::installed.genomes()), 
    value = TRUE
  )
  
  if( length(genome) == 0 ){
    
    cat("\nInstalled genomes include:")
    print(unique(BSgenome::installed.genomes()))
    cat("\nSelected reference genome not in list.")
    stop("Error: Genome not available.")
    
  }else if( length(genome) > 1 ){
    
    cat("\nInstalled genomes include:")
    print(unique(BSgenome::installed.genomes()))
    cat(
      "\nPlease be more specific about reference genome.",
      "Multiple matches to input."
    )
    stop("Error: Multiple genomes requested.")
    
  }
  
  suppressMessages(library(genome, character.only = TRUE))
  
  ref_genome <- get(genome)
  
}

## Get versioning and params

soft_version <- eval_data$params$soft_version

build_version <- eval_data$params$build_version

signature <- paste(
  unique(sort(unlist(lapply(configs, "[[", "signature")))), 
  collapse = ", ")

## Determine processing parameters
## Some parameters will need to be an "all or nothing" approach, including:
##   - UMItags
##   - recoverMultihits
## Depending on these parameters others (upstream/downstream_dist, ...) may need
## to be consistent between runs otherwise, the primary config file (first one),
## will be used for parameterization.

umitag_option <- all(unlist(lapply(configs, "[[", "UMItags")))
multihit_option <- all(unlist(lapply(configs, "[[", "recoverMultihits")))

abundance_option <- unique(
  tolower(unlist(lapply(configs, "[[", "Abundance_Method")))
)[1]

if( is.na(abundance_option) ) abundance_option <- "Fragment"

if( abundance_option == "umi" & !umitag_option ){
  stop(
    "\n  Abundance method has been set to use UMItags, yet the current",
    "\n  configuration does not capture UMItag data (UMItags : FALSE).",
    "\n  Please correct this inconsistency before continuing analysis."
  )
}

if( multihit_option ){
  
  upstream_dist <- unique(sapply(configs, function(x) x$upstreamDist))
  downstream_dist <- unique(sapply(configs, function(x) x$downstreamDist))
  pile_up_min <- unique(sapply(configs, function(x) x$pileUpMin))
  
  if( 
    length(upstream_dist) > 1 | 
    length(downstream_dist) > 1 | 
    length(pile_up_min) > 1 
  ){
    
    stop(
      "\n  Inconsistant upstream or downstream distances between config files.\n",
      "  Comparisons between groups with different run specific criteria\n", 
      "  is not recommended when considering the recover multihit option.\n"
    )
    
  }
  
}else{
  
  upstream_dist <- configs[[1]]$upstreamDist
  downstream_dist <- configs[[1]]$downstreamDist
  pile_up_min <- configs[[1]]$pileUpMin
  
}

## Combine sampleInfo files

sample_info <- eval_data$spec_info$sample_info

specimen_levels <- eval_data$params$specimen_levels
alt_specimen_levels <- eval_data$params$alt_specimen_levels

support_present <- nrow(eval_data$spec_info$supp_data) > 0

## Identify all targets used
target_tbl <- eval_data$spec_info$target_tbl %>%
  dplyr::select(-run_set) %>%
  dplyr::distinct() %>%
  dplyr::rename(
    "Nuclease" = nuclease,
    "Target Name" = target,
    "Sequence" = sequence
  )

## Identify on-target edit sites
on_targets <- eval_data$spec_info$on_targets

## Treatment across runs
treatment <- eval_data$spec_info$treatment
treatment_df <- eval_data$spec_info$treatment_df

## Nuclease profiles
nuc_profiles <- eval_data$spec_info$nuclease_profiles

## Combo information
nuc_treatment_unmod_df <- eval_data$spec_info$nuclease_df %>%
  dplyr::full_join(treatment_df, by = c("run_set", "specimen"))

nuclease_treatment_df <- eval_data$spec_info$nuclease_treatment_df
combos_set_tbl <- eval_data$spec_info$combos_set_tbl

combos_tbl <- combos_set_tbl %>%
  dplyr::select(-run_set) %>%
  dplyr::distinct() %>%
  dplyr::mutate(combo = paste0("(", combo, ")")) %>%
  dplyr::rename(
    "Combination" = combo,
    "Nuclease" = nuclease,
    "Treatment" = treatment
  )

## Load in supporting information ----
supp_data <- eval_data$spec_info$supp_data

## Consolidate supplementary data ----
spec_overview <- eval_data$spec_info$spec_overview %>%
  dplyr::mutate(
    specimen = ifelse(
      nuc_treatment_unmod_df$nuclease[
        match(specimen, nuc_treatment_unmod_df$specimen)
      ] == "Mock" | nuc_treatment_unmod_df$treatment[
        match(specimen, nuc_treatment_unmod_df$specimen)
      ] == "Mock",
      as.character(specimen),
      as.character(nuclease_treatment_df$alt_specimen)[
        match(specimen, nuclease_treatment_df$specimen)
      ]
    )
  )

if( length(unique(spec_overview$run_set)) == 1 ){
  spec_overview <- dplyr::select(spec_overview, -run_set)
}else{
  spec_overview <- dplyr::rename(spec_overview, "Run Name" = run_set)
}

annot_overview <- eval_data$spec_info$annot_overview

combo_overview <- nuclease_treatment_df %>%
  dplyr::left_join(annot_overview, by = "specimen") %>%
  dplyr::mutate(
    annotation = paste0(as.character(annotation), " (", combo, ")"),
    annotation = factor(annotation, levels = unique(annotation))
  )


## Read in experimental data and contatenate different sets ----
incorp_data <- eval_data$incorp_data


## Info graphic data ----
graphic_order <- c("algnmts", "pile_up_algns", "paired_algns", "matched_algns")
graphic_data <- incorp_data[graphic_order]

graphic_grl <- GenomicRanges::GRangesList(lapply(
  graphic_data, 
  GenomicRanges::makeGRangesFromDataFrame, 
  seqinfo = GenomicRanges::seqinfo(ref_genome)
))


# Genomic Distribution of edited sites ----
genomic_grl <- GenomicRanges::GRangesList(lapply(
  graphic_data, 
  function(x){
    
    y <- makeGRangesFromDataFrame(x, seqinfo = seqinfo(ref_genome))
    mcols(y) <- combo_overview[
      match(x$alt_specimen, combo_overview$alt_specimen), 
      "annotation", 
      drop = FALSE
    ]
    
    y
    
  }
))

num_conds <- max(length(unique(combo_overview$annotation)), 1)

names(genomic_grl) <- c(
  "All Align.", "Pileup Align.", "Flanking Pairs", "Target Matched"
)


# On-target summary ----
ot_tbl_summary <- eval_data$summary_tbls$ot_tbl_summary %>%
  dplyr::rename("specimen" = alt_specimen) %>%
  dplyr::mutate(
    annotation = stringr::str_remove(annotation, "\\([\\w]+\\)$")
  ) %>%
  dplyr::select(
    "specimen", "annotation", "ot_algns_pct", "ot_pile_pct", 
    "ot_pair_pct", "ot_match_pct"
  )

ot_eff_summary <- eval_data$summary_tbls$ot_eff_summary %>%
  dplyr::rename("specimen" = alt_specimen) %>%
  dplyr::mutate(
    annotation = stringr::str_remove(annotation, "\\([\\w]+\\)$")
  ) %>%
  dplyr::select(specimen, annotation, dplyr::everything())

eval_summary <- eval_data$summary_tbls$eval_summary

# On-target distribution of incorporations ----
on_tar_dists <- eval_data$edit_models$on_tar_dists
sites_included <- eval_data$edit_models$sites_included

# Off-target summary ----
ft_tbl_summary <- eval_data$summary_tbls$ft_tbl_summary %>%
  dplyr::rename("specimen" = alt_specimen) %>%
  dplyr::mutate(
    annotation = stringr::str_remove(annotation, "\\([\\w]+\\)$")
  ) %>%
  dplyr::select(
    "specimen", "annotation", "ft_algns", "ft_pile", "ft_pair", "ft_match"
  )


# Off-target sequence analysis ----
ft_seqs_list <- eval_data$ft_data


# Onco-gene enrichment analysis ----
enrich_df <- eval_data$enrich_data$enrich_df


# Data passed to Rmd for report generation ----
set_names <- eval_data$params$set_names

# Normalize file output path
write(c(), file = args$output)
args$output <- normalizePath(args$output)
unlink(args$output)

output_path <- unlist(strsplit(args$output, "/"))
output_dir <- paste(output_path[seq_len(length(output_path)-1)], collapse = "/")
output_file <- output_path[length(output_path)]

if( args$format == "html" & !stringr::str_detect(output_file, ".html$") ){
  output_file <- paste0(output_file, ".html")
}

if( args$format == "pdf" & !stringr::str_detect(output_file, ".pdf$") ){
  output_file <- paste0(output_file, ".pdf")
}

figure_path <- file.path(
  output_dir, gsub("[\\w]+$", "figures", output_file, perl = TRUE)
)

null <- dir.create(figure_path)

if( args$tables ){
  
  tables_path <- file.path(
    output_dir, gsub("[\\w]+$", "tables", output_file, perl = TRUE)
  )
  
  null <- dir.create(tables_path)
  
}

if( args$data ){

  if( args$format == "html" ){
    
    save.image(file = file.path(
      output_dir, stringr::str_replace(output_file, ".html$", ".RData")
    )) 
    
  }else if( args$format == "pdf" ){
    
    save.image(file = file.path(
      output_dir, stringr::str_replace(output_file, ".pdf$", ".RData")
    )) 
    
  }
  
}

if( args$format == "html" ){
  
  css_path <- normalizePath(
    file.path(root_dir, "tools/rscripts/report_templates/iguide.css")
  )
  
  rmarkdown::render(
    input = template_path,
    output_format = output_format, 
    output_file = output_file,
    output_dir = output_dir,
    output_options = list("css" = css_path)
  )
  
}else{
  
  rmarkdown::render(
    input = template_path,
    output_format = output_format, 
    output_file = output_file,
    output_dir = output_dir
  )
  
}

if( !args$figures ){
  
  tmp_fig_paths <- c(
    list.files(
      path = figure_path, pattern = "incorp_dist", full.names = TRUE
    ),
    list.files(
      path = figure_path, pattern = "genomic_dens", full.names = TRUE
    ),
    list.files(
      path = figure_path, pattern = "off_target_seqs", full.names = TRUE
    )
  )
  
  cat(sprintf("Removing temporary files: %s\n", tmp_fig_paths), sep = "")
  null <- file.remove(tmp_fig_paths)
  cat("Removing temorary directory:", figure_path, "\n")
  null <- file.remove(figure_path)
  
}

q()
