#' For anyone reviewing the code below, the following is a small style guide 
#' outlining the various formats for the code. 
#' 
#' Names with "_": objects, inlucding data.frames, GRanges, vectors, ...
#' Names in caMel format: functions or components of objects (i.e. columns 
#' within a data.frame).
#' Names with ".": arguments / options for functions

options(stringsAsFactors = FALSE, scipen = 99, width = 180)
set.seed(1)


# Set up and gather command line arguments ----
parser <- argparse::ArgumentParser(
  description = "Generate an iGUIDE summary from input evaluation data.",
  usage = paste(
    "Rscript generate_IGUIDE_summary.R <eval.rds> [optional args]",
    "[-h/--help, -v/--version]"
  )
)

parser$add_argument(
  "eval", nargs = 1, type = "character",
  help = paste(
    "Evaluation dataset, in rds format. Can be generated by the",
    "'iguide eval' subcommand."
  )
)

parser$add_argument(
  "-o", "--output", nargs = 1, type = "character", default = FALSE,
  help = paste(
    "Output report file, extention not required. Will be writen as text",
    "file. If no output given, results will be printed to screen.",
    "Example output name: summary.iGUIDE_Run.txt or summary.iguide_run"
  )
)

parser$add_argument(
  "-p", "--power_filt", nargs = 1, type = "integer", default = 0,
  help = paste(
    "Specify a integer between 0 and 100 indicating the percent for",
    "which to filter statistical comparisons for gene enrichment based",
    "on power."
  ),
  metavar = "INT"
)

parser$add_argument(
  "-m", "--mesl_filt", nargs = 1, type = "integer", default = 0,
  help = paste(
    "Specify a integer between 0 and 100 indicating the likelyhood for",
    "which to filter off-target sites based on the Mean Edit Site",
    "Likelyhood (MESL) model."
  ),
  metavar = "INT"
)

parser$add_argument(
  "--iguide_dir", nargs = 1, type = "character", default = "IGUIDE_DIR",
  help = "iGUIDE install directory path, do not change for normal applications."
)


args <- parser$parse_args(commandArgs(trailingOnly = TRUE))

if( !dir.exists(args$iguide_dir) ){
  root_dir <- Sys.getenv(args$iguide_dir)
}else{
  root_dir <- args$iguide_dir
}

if( !dir.exists(root_dir) ){
  stop(paste0("\n  Cannot find install path to iGUIDE: ", root_dir, ".\n"))
}else{
  args$iguide_dir <- root_dir
}

# Normalize file output path
if( args$output != FALSE ){
  
  write(c(), file = args$output)
  args$output <- normalizePath(args$output)
  unlink(args$output)
  
  output_path <- unlist(strsplit(args$output, "/"))
  output_dir <- paste(output_path[seq_len(length(output_path)-1)], collapse = "/")
  output_file <- output_path[length(output_path)]
  
  if( !stringr::str_detect(output_file, ".txt$") ){
    output_file <- paste0(output_file, ".txt")
  }
  
  args$output <- file.path(output_dir, output_file)
  unlink(args$output)

}

## Additional functions to help with output 
catOrWrite <- function(obj, args, big.mark = ",", style = "simple", ...){
  
  if( is.data.frame(obj) ){
    
    obj <- pander::pandoc.table.return(
      obj, row.names = FALSE, 
      split.cells = Inf, split.tables = Inf, style = style, 
      justify = ifelse(
        sapply(seq_len(ncol(obj)), function(i){
          is.numeric(obj[, i, drop = TRUE])
        }),
        "right", "centre"
      ),
      plain.ascii = TRUE,
      ...
    )
    
    obj <- stringr::str_remove(obj, "\n")

  }else{
    
    obj <- paste0(obj, "\n")
    
  }

  if( args$output != FALSE ){
    cat(obj, file = args$output, append = TRUE)
  }else{
    cat(obj)
  }

}


# Load dependancies ----
add_packs <- c("magrittr", "iguideSupport")

add_packs_loaded <- suppressMessages(
  sapply(add_packs, require, character.only = TRUE)
)

if( !all(add_packs_loaded) ){
  
  print(
    data.frame(
      "R-Packages" = names(add_packs_loaded), 
      "Loaded" = add_packs_loaded
    ), 
    right = FALSE,
    row.names = FALSE
  )
  
  stop("Check dependancies.\n")
  
}


# Import metadata and consolidate into report objects ----
if( file.exists(file.path(root_dir, args$eval)) ){
  
  eval_path <- normalizePath(file.path(root_dir, args$eval))
  
}else if( file.exists(file.path(args$eval)) ){
  
  eval_path <- normalizePath(file.path(args$eval))
  
}else{
  
  stop("\n  Cannot find evaluation dataset: ", args$eval, "\n")
  
}

eval_data <- readRDS(eval_path)

## Configuration
configs <- eval_data$params$configs

## Get versioning and params
set_names <- eval_data$params$set_names
soft_version <- eval_data$params$soft_version
build_version <- eval_data$params$build_version

signature <- paste(
  unique(sort(unlist(lapply(configs, "[[", "signature")))), 
  collapse = ", ")

umitag_option <- all(unlist(lapply(configs, "[[", "UMItags")))
abundance_option <- unique(
  tolower(unlist(lapply(configs, "[[", "Abundance_Method")))
)[1]

if( is.na(abundance_option) ) abundance_option <- "Fragment"

if( tolower(abundance_option) == "umi" & !umitag_option ){
  stop(
    "\n  Abundance method has been set to use UMItags, yet the current",
    "\n  configuration does not capture UMItag data (UMItags : FALSE).",
    "\n  Please correct this inconsistency before continuing analysis."
  )
}


# Start summary ----
null <- catOrWrite(
  paste0(
    "iGUIDE Summary:\n",
    "  run(s) : ", set_names, "\n",
    "  author(s): ", signature, "\n",
    "  generated : ", "[", paste(Sys.time()), "]\n",
    "  software version : ", soft_version, "\n",
    "  build version : ", build_version, "\n",
    "\n************************************************************\n"
  ), 
  args
)
 
# Analysis overview table ----
eval_summary <- eval_data$summary_tbls$eval_summary %>%
  dplyr::mutate(Specimen = stringr::str_remove(Specimen, "\\([\\w]+\\)$"))

eval_summary[is.na(eval_summary)] <- 0

null <- catOrWrite(
  "Table 1. Analysis overview with specific data highlights.", 
  args
)

null <- catOrWrite(eval_summary, args, missing = 0, style = "multiline")

null <- catOrWrite("", args)

# Specimen summary table ----
specimen_levels <- eval_data$params$specimen_levels
alt_specimen_levels <- eval_data$params$alt_specimen_levels
spec_overview <- eval_data$spec_info$spec_overview
on_targets <- eval_data$spec_info$on_targets

if( length(unique(spec_overview$run_set)) == 1 ){
  spec_overview <- dplyr::select(spec_overview, -run_set)
}else{
  spec_overview <- dplyr::rename(spec_overview, "Run Name" = run_set)
}

spec_overview <- eval_data$incorp_data$algnmts %>%
  dplyr::mutate(type = ifelse(
    is.na(edit.site),
    "Independent", 
    ifelse(edit.site %in% expandPosStr(on_targets), "On-target", "Off-target")
    )
  ) %>%
  dplyr::group_by(alt_specimen, type) %>%
  dplyr::summarise(abund = sum(abund)) %>%
  dplyr::ungroup() %>%
  dplyr::mutate(
    alt_specimen = factor(alt_specimen, levels = alt_specimen_levels),
    type = factor(type, levels = c("On-target", "Off-target", "Independent"))
  ) %>%
  tidyr::spread(key = "type", value = "abund", fill = 0) %>%
  tidyr::complete(
    alt_specimen, 
    fill = list("On-target" = 0, "Off-target" = 0, "Independent" = 0)
  ) %>%
  dplyr::mutate(
    specimen = factor(
      stringr::str_remove(as.character(alt_specimen), "\\([\\w]+\\)$"), 
      levels = specimen_levels
    )
  ) %>%
  dplyr::left_join(spec_overview, ., by = "specimen") %>%
  dplyr::select(alt_specimen, dplyr::everything(), -specimen) %>%
  dplyr::rename(specimen = alt_specimen)

null <- catOrWrite(
  "Table 2. Specimen overview covering reads, umitags, and alignments:", 
  args
)

null <- catOrWrite(spec_overview, args, missing = 0, style = "multiline")

null <- catOrWrite("", args)

# Combination info table ----
combo_overview <- eval_data$spec_info$combo_overview

combos_tbl <- eval_data$spec_info$combos_set_tbl %>%
  dplyr::select(-run_set) %>%
  dplyr::distinct() %>%
  dplyr::mutate(combo = paste0("(", combo, ")")) %>%
  dplyr::rename(
    "Combination" = combo,
    "Nuclease" = nuclease,
    "Treatment" = treatment
  )

null <- catOrWrite(
  "Table 3. Combinations of nuclease(s) and treatment(s).",
  args
)

null <- catOrWrite(combos_tbl, args)

# Target info table ----
## Identify all targets used
target_tbl <- eval_data$spec_info$target_tbl %>%
  dplyr::select(-run_set) %>%
  dplyr::distinct() %>%
  dplyr::rename(
    "Nuclease" = nuclease,
    "Target Name" = target,
    "Sequence" = sequence
  ) %>%
  dplyr::mutate(
    "Edit Loci" = sapply(
      `Target Name`, 
      function(x){
        paste(on_targets[which(names(on_targets) == x)], collapse = "\n")
      }
    )
  )

null <- catOrWrite(
  "Table 4. Target pattern table specifying sequences and edited loci:", 
  args
)

null <- catOrWrite(
  target_tbl, args, style = "multiline", keep.line.breaks = TRUE
)

null <- catOrWrite("", args)


# On-target summary table ----
ot_tbl_summary <- eval_data$summary_tbls$ot_tbl_summary %>%
  dplyr::mutate(
    specimen = combo_overview$specimen[
      match(alt_specimen, combo_overview$alt_specimen)
      ]
  ) %>%
  dplyr::select(
    "specimen", "annotation", "ot_algns_pct", "ot_pile_pct", 
    "ot_pair_pct", "ot_match_pct"
  )

names(ot_tbl_summary) <- c(
  "Specimen", "Annotation", "All\nAlign.", "Align.\nPileups", 
  "Flanking\nPairs", "Target\nMatched"
)

null <- catOrWrite(
  "Table 5. On-target editing percentages based on alignment criteria:",
  args
)

null <- catOrWrite(
  ot_tbl_summary, args, 
  digits = 4, round = 2, style = "multiline", missing = 0, 
  keep.line.breaks = TRUE
)

null <- catOrWrite("", args)


# On-target distribution of incorporations ----
on_tar_dists <- eval_data$edit_models$on_tar_dists
sites_included <- eval_data$edit_models$sites_included

on_tar_dist_summary <- on_tar_dists %>%
  dplyr::group_by(annotation, target) %>%
  dplyr::summarise(
    quant = paste(
      round(quantile(S4Vectors::Rle(abs(edit.site.dist), cnt)), digits = 0), 
      collapse = ";"
    )
  ) %>%
  dplyr::ungroup() %>%
  dplyr::left_join(
    dplyr::select(sites_included, annotation, target, prop),
    by = c("annotation", "target")
  ) %>%
  dplyr::select(annotation, target, prop, quant) %>%
  tidyr::separate(
    quant, paste0(as.character(100*seq(0, 1, 0.25)), "%"), sep = ";"
  ) %>%
  dplyr::rename(
    "Annotation" = annotation,
    "Target" = target,
    "Inclusion Pct." = prop
  )

null <- catOrWrite(
  "Table 6. On-target incorporation profile, quantile counts given in % columns:",
  args
)

null <- catOrWrite(on_tar_dist_summary, args, style = "multiline", missing = 0)

null <- catOrWrite("", args)

# On-target editing efficiency ----
ot_eff_summary <- eval_data$summary_tbls$ot_eff_summary %>%
  dplyr::mutate(
    specimen = combo_overview$specimen[
      match(alt_specimen, combo_overview$alt_specimen)
      ]
  ) %>%
  dplyr::select(specimen, dplyr::everything(), -alt_specimen)

names(ot_eff_summary)[c(1,2)] <- c("Specimen", "Annotation")

null <- catOrWrite(
  "Table 7. Estimate of On-target editing efficiency (percent) for each target by specimen.",
  args
)

null <- catOrWrite(ot_eff_summary, args, style = "multiline", missing = "-")

null <- catOrWrite("", args)

# Off-target summary ----
ft_tbl_summary <- eval_data$summary_tbls$ft_tbl_summary %>%
  dplyr::mutate(
    specimen = combo_overview$specimen[
      match(alt_specimen, combo_overview$alt_specimen)
      ]
  ) %>%
  dplyr::select(
    "specimen", "annotation", "ft_algns", "ft_pile", "ft_pair", "ft_match"
  )

names(ft_tbl_summary) <- c(
  "Specimen", "Annotation", "All\nAlign.", "Align.\nPileups", 
  "Flanking\nPairs", "Target\nMatched"
)

null <- catOrWrite(
  "Table 8. Off-target loci counts from criteria-based alignments:",
  args
)

null <- catOrWrite(
  ft_tbl_summary, args,
  digits = 1, big.mark = ",", missing = 0, 
  keep.line.breaks = TRUE, style = "multiline"
)

null <- catOrWrite("", args)


# Onco-gene enrichment analysis ----
enrich_df <- eval_data$enrich_data$enrich_df %>%
  dplyr::select(
    origin, annotation, total, 
    onco, onco.p.value, onco.power, 
    special, special.p.value, special.power
  ) %>%
  dplyr::filter(
    onco.power >= args$power_filt / 100 | special.power >= args$power_filt / 100
  )


names(enrich_df) <- c(
  "Origin", "Annotation", "Total Gene Count", "Onco Related Count", 
  "Onco Enrich. p-value", "Onco Test Power", "Special Gene Count", 
  "Special Enrich. p-value", "Special Test Power"
)

null <- catOrWrite(
  "Table 9. Off-target gene enrichment:",
  args
)

if( nrow(enrich_df) > 0 ){
  
  names(enrich_df) <- gsub(" ", "\n", names(enrich_df))
  enriched_idx <- which(enrich_df <= 0.05, arr.ind = TRUE)
  enriched_idx <- enriched_idx[enriched_idx[,2] >= 6, , drop = FALSE]
  enrich_df[,5] <- sprintf("%.3f", round(enrich_df[,5], digits = 3))
  enrich_df[,6] <- sprintf("%.3f", round(enrich_df[,6], digits = 3))
  enrich_df[,8] <- sprintf("%.3f", round(enrich_df[,8], digits = 3))
  enrich_df[,9] <- sprintf("%.3f", round(enrich_df[,9], digits = 3))
  
  null <- catOrWrite(
    enrich_df, args,
    digits = 4, style = "multiline", keep.line.breaks = TRUE
  )
  
}else if( args$power_filt > 0){
  
  null <- catOrWrite(
    paste0(
      "  No gene enrichment was observed with at a power ",
      "greater than or equal to ", args$power_filt, "%."
    ), 
    args
  )
  
}else{
  
  null <- catOrWrite("  No gene enrichment was observed.", args)
  
}

null <- catOrWrite("", args)

# Off-target sequence analysis ----
nuc_profiles <- eval_data$spec_info$nuclease_profiles
ft_seqs_list <- eval_data$ft_data

full_target_seqs <- structure(
  sapply(seq_len(nrow(target_tbl)), function(i){
    
    nuc <- target_tbl$Nuclease[i]
    sequence <- target_tbl$Sequence[i]
    
    ifelse(
      nuc_profiles[[nuc]]$PAM_Loc == "3p",
      paste0(sequence, nuc_profiles[[nuc]]$PAM),
      ifelse(
        nuc_profiles[[nuc]]$PAM_Loc == "5p",
        paste0(nuc_profiles[[nuc]]$PAM, sequence),
        sequence
      )
    )
    
  }),
  names = target_tbl$`Target Name`
)

null <- lapply(seq_along(ft_seqs_list), function(i){
  
  null <- catOrWrite(paste0("Table ", i+9, ". Off-Target Loci:"), args)
  null <- catOrWrite(paste0("  Annotation : ", names(ft_seqs_list)[i]), args)
  
  target_ref_seq <- full_target_seqs[unique(ft_seqs_list[[i]]$target.seq)]
  
  null <- dplyr::select(
      ft_seqs_list[[i]], target, gene_id, edit.site, abund, MESL, 
      aligned.sequence, mismatch
    ) %>%
    dplyr::mutate(
      aligned.sequence = divSeq(aligned.sequence, target_ref_seq)
    ) %>%
    dplyr::rename(
      "Target" = target, "Gene ID" = gene_id, "Edit Site" = edit.site,
      "Abund." = abund, "Aligned Sequence" = aligned.sequence,
      "Mismatch" = mismatch
    ) %>%
    dplyr::filter(MESL >= args$mesl_filt) %>%
    catOrWrite(args)
  
  null <- catOrWrite("", args)
  
})

q()
